{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "52d8da7d-c53d-4975-b203-b6ea9abdb85b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 200)\n",
    "\n",
    "rd = pd.read_excel('./Resources/combined_phecodex_list.xlsx')\n",
    "cd = rd.loc[rd['Type'] == 'Common']\n",
    "\n",
    "# Optional plotting functions\n",
    "if True:\n",
    "    import re\n",
    "    import os\n",
    "    import matplotlib\n",
    "    import matplotlib.font_manager as font_manager\n",
    "\n",
    "    font_path = '/sc/arion/projects/GENECAD/Robert/arial.ttf'\n",
    "    assert os.path.exists(font_path)\n",
    "    font_manager.fontManager.addfont(font_path)\n",
    "    prop = font_manager.FontProperties(fname=font_path)\n",
    "\n",
    "    matplotlib.rc('font', family='sans-serif')\n",
    "    matplotlib.rcParams.update({\n",
    "        'font.size': 12,\n",
    "        'font.sans-serif': prop.get_name(),\n",
    "        'axes.titlesize': 12\n",
    "    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a0c40f3a-9115-4054-b301-368fa134d3eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "alldata = pd.read_pickle('./Final/ml_input.pkl')\n",
    "\n",
    "clin = ['clingen','eva','gene2phenotype','genomics_england','orphanet','uniprot_literature', 'uniprot_variants']\n",
    "ot = ['clin_ot','hgmd','gwas_credible_sets','expression_atlas','impc','europepmc']\n",
    "mantis = ['mantis']\n",
    "cc = ['cc_common_max_p','cc_rare_max_p','cc_rare_burden_max_p','cc_ultrarare_max_p']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30adbf15-7f8f-449b-930d-7e211b5f9d93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Additional labels for dataset (optional) \n",
    "\n",
    "truemoa = alldata[['gene','phecode','moa']].dropna()\n",
    "truemoa['moa'] = truemoa['moa'].astype(str)\n",
    "truemoa.loc[truemoa['moa'].str.contains('Negative'), 'moa_clean'] = 'Inhibition'\n",
    "truemoa.loc[truemoa['moa'].str.contains('Positive'), 'moa_clean'] = 'Activation'\n",
    "truemoa.loc[truemoa['moa'].str.contains('Other'), 'moa_clean'] = 'Other or unclear'\n",
    "truemoa.loc[(truemoa['moa'].str.contains('Positive') & truemoa['moa'].str.contains('Negative')), 'moa_clean'] = 'Other or unclear'\n",
    "truemoa = truemoa[['gene','phecode','moa_clean']]\n",
    "\n",
    "dtp = pd.read_pickle('./Final/drug_target_phase.pkl')\n",
    "names = pd.read_pickle('./Resources/targets_with_names_cleaned.pkl')[['chembl_id','dc_name']].dropna().drop_duplicates('chembl_id').rename({'dc_name':'name'},axis=1)\n",
    "dtp = dtp.merge(names, how='left')\n",
    "\n",
    "moa = dtp.groupby(['phecode','gene'])['moa'].value_counts().reset_index()\n",
    "moa = pd.pivot_table(moa, index=['phecode','gene'], columns='moa', values='count').reset_index()\n",
    "moa = moa.fillna(0)\n",
    "\n",
    "moa.loc[(moa['Positive modulator'] > moa['Negative modulator']) & (moa['Positive modulator'] >= moa['Other/unknown']), 'moa'] = 'Activation'\n",
    "moa.loc[(moa['Negative modulator'] > moa['Positive modulator']) & (moa['Negative modulator'] >= moa['Other/unknown']), 'moa'] = 'Inhibition'\n",
    "moa.loc[moa['Positive modulator'] == moa['Negative modulator'], 'moa'] = 'Conflicting'\n",
    "moa.loc[(moa['Positive modulator'] == 0) & (moa['Negative modulator'] == 0), 'moa'] = 'Other or unknown'\n",
    "moa.loc[(moa['moa'].isna()) & (moa['Negative modulator'] > 1), 'moa'] = 'Inhibition' \n",
    "moa.loc[(moa['moa'].isna()) & (moa['Positive modulator'] > 1), 'moa'] = 'Activation' \n",
    "moa['moa'] = moa['moa'].fillna('Other or unknown')\n",
    "moa = moa[['phecode','gene','moa']].drop_duplicates()\n",
    "\n",
    "ind = dtp[['gene','phecode','phase','name']].drop_duplicates()\n",
    "ind = ind.groupby(['gene','phecode','name'])['phase'].max().reset_index()\n",
    "ind['phase'] = ind['phase'].map({0.5:'preclinical',1:'I',2:'II',3:'III',4:'IV'})\n",
    "ind['label'] = ind['name'] + ' (' + ind['phase'].astype(str) + ')'\n",
    "ind = ind.groupby(['gene','phecode'])['label'].unique().reset_index()\n",
    "ind['label'] = ind['label'].astype(str).str.replace(\" 'Unknown'\",\"\").str.replace(\"' '\",\", \").str.replace(\"'\",\"\").str.replace(\"[\",\"\").str.replace(\"]\",\"\")\n",
    "\n",
    "tn = pd.read_pickle('./Resources/targets_with_names_cleaned.pkl').sort_values(['gene','dc_name'])\n",
    "tn = tn.groupby(['gene','moa'])['dc_name'].unique().reset_index()\n",
    "tn['dc_name'] = tn['dc_name'].astype(str)\n",
    "tn['dc_name'] = tn['dc_name'].str.replace(\"' '\",\", \",regex=False).str.replace(\"['\",\"\",regex=False).str.replace(\"']\",\"\",regex=False)\n",
    "tn = pd.pivot(tn, index='gene', columns='moa', values='dc_name').reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cd1cf965-4c64-45eb-be86-a8a806348f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "pred = pd.read_pickle('./GPS/Main/Predictions/apc_predictions_ot_mantis_cc.pkl').sort_values('prediction', ascending=False)\n",
    "pred = pred.merge(rd[['phecode','phecode_string']])\n",
    "pred = pred.merge(alldata[['id','drug_gene','Tclin']])\n",
    "pred = pred.merge(moa, how='left')\n",
    "pred = pred.merge(tn, on='gene', how='left')\n",
    "pred = pred.merge(ind, how='left')\n",
    "pred.loc[(pred['indication'] == 1) & (pred['label'].isna()), 'label'] = 'unknown'\n",
    "\n",
    "pred['Positive modulator'] = pred['Positive modulator'].str.replace(\"' '\",\", \",regex=False).str.replace(\"'\\n '\",\", \",regex=False)\n",
    "pred['Negative modulator'] = pred['Negative modulator'].str.replace(\"' '\",\", \",regex=False).str.replace(\"'\\n '\",\", \",regex=False)\n",
    "pred['Other'] = pred['Other'].str.replace(\"' '\",\", \",regex=False).str.replace(\"'\\n '\",\", \",regex=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "801d179b-54a1-4b11-ad3f-bb51280fa5da",
   "metadata": {},
   "outputs": [],
   "source": [
    "main = pred[['gene','phecode','phecode_string','prediction','indication','phase',\n",
    "         'moa','label',\n",
    "         'drug_gene','Tclin',\n",
    "         'Positive modulator','Negative modulator','Other']]\n",
    "main['phecode'] = main['phecode'] + \" [\" + main['phecode_string'] + ']'\n",
    "main = main.drop('phecode_string',axis=1)\n",
    "\n",
    "scores = main.set_axis(['Gene','Phecode','RareGPS','Indication','Max phase',\n",
    "                      'Indicated mechanism','Indicated drugs',\n",
    "                      'Druggable gene','DrugnomeAI',\n",
    "                      'Activators','Inhibitors','Other drugs'],axis=1)\n",
    "\n",
    "scores_top = scores.loc[scores['RareGPS'] >= 0.036805]\n",
    "\n",
    "scores[['RareGPS','DrugnomeAI']] = scores[['RareGPS','DrugnomeAI']].round(3)\n",
    "scores_top[['RareGPS','DrugnomeAI']] = scores_top[['RareGPS','DrugnomeAI']].round(3)\n",
    "\n",
    "scores.to_csv('./Public/RareGPS_all.csv', index=False)\n",
    "scores_top.to_csv('./Public/RareGPS_top.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8dbc4236-9584-42be-b443-4db67a655657",
   "metadata": {},
   "outputs": [],
   "source": [
    "evidence = alldata[['gene','phecode','expression_atlas','hgmd','gwas_credible_sets','impc',\n",
    "                    'clin_ot','europepmc','cc_common_max_p','cc_rare_max_p','cc_rare_burden_max_p',\n",
    "                    'cc_ultrarare_max_p','mantis']]\n",
    "\n",
    "evidence = pred[['gene','phecode','phecode_string','prediction','indication','phase']].merge(evidence, on=['gene','phecode'])\n",
    "evidence['phecode'] = evidence['phecode'] + \" [\" + evidence['phecode_string'] + ']'\n",
    "evidence = evidence.drop('phecode_string',axis=1)\n",
    "\n",
    "evidence['categories'] = 0\n",
    "evidence.loc[evidence[['clin_ot','hgmd']].max(axis=1) > 0, 'categories'] += 1\n",
    "evidence.loc[(evidence[['cc_common_max_p','cc_rare_max_p',\n",
    "                       'cc_rare_burden_max_p','cc_ultrarare_max_p']].max(axis=1) > 1.301) | (evidence['gwas_credible_sets'] > 0), 'categories'] += 1\n",
    "evidence.loc[evidence[['europepmc']].max(axis=1) > 0, 'categories'] += 1\n",
    "evidence.loc[evidence[['mantis']].max(axis=1) > 0, 'categories'] += 1\n",
    "evidence.loc[evidence[['impc']].max(axis=1) > 0, 'categories'] += 1\n",
    "evidence.loc[evidence[['expression_atlas']].max(axis=1) > 0, 'categories'] += 1\n",
    "\n",
    "evidence = evidence.set_axis(['Gene','Phecode','RareGPS','Indication','Max phase','Gene expression',\n",
    "                              'HGMD','L2G','Mouse models','OTP clinical genetics','Text mining',\n",
    "                              'Common variants','Rare variants','Rare variants (gene-level)','Ultrarare variants (gene-level)',\n",
    "                              'Mantis-ML','Sources'],axis=1)\n",
    "evidence[['RareGPS','Gene expression','HGMD','L2G','Mouse models','OTP clinical genetics','Text mining',\n",
    "          'Common variants','Rare variants','Rare variants (gene-level)','Ultrarare variants (gene-level)','Mantis-ML']] = evidence[['RareGPS','Gene expression','HGMD','L2G','Mouse models','OTP clinical genetics','Text mining',\n",
    "          'Common variants','Rare variants','Rare variants (gene-level)','Ultrarare variants (gene-level)','Mantis-ML']].round(3)\n",
    "\n",
    "evidence.to_csv('./Public/RareGPS_evidence.csv', index=False)\n",
    "\n",
    "evidence_top = evidence.loc[evidence['RareGPS'] >= 0.036805]\n",
    "evidence.to_csv('./Public/RareGPS_top_evidence.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5e953fde-17f0-4490-ab71-2a9b4c85eaa2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sources\n",
       "1    2004186\n",
       "2     517550\n",
       "0     453034\n",
       "3      41803\n",
       "4       4694\n",
       "5        658\n",
       "6         40\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "evidence['Sources'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab1c8318-95c4-46f8-b842-183ee1ab6c84",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
